A04_attacks-and-mitigations.md

```diff
@@ -1335,21 +1336,21 @@ on the client policy or the grant associated with the refresh token
 Resource servers may make access control decisions based on the identity of a
 resource owner for which an access token was issued, or based on the identity
 of a client in the client credentials grant. If both options are possible,
-depending on the details of the implementation, a client's identity may be
-mistaken for the identity of a resource owner. For example, if a client is able
+depending on the details of the implementation, a resource server may mistake
+a client's identity for the identity of a resource owner. For example, if a client is able
 to choose its own `client_id` during registration with the authorization server,
 a malicious client may set it to a value identifying an end-user (e.g., a `sub`
 value if OpenID Connect is used). If the resource server cannot properly
-distinguish between access tokens issued to clients and access tokens issued to
-end-users, the client may then be able to access resource of the end-user.
+distinguish between access tokens linked to client resources and access tokens linked to
+end-user resources, the client may then be able to access resources belonging to the end-user.
 
 
 ### Countermeasures {#client_impersonating_countermeasures}
 
 If the authorization server has a common namespace for client IDs and user
-identifiers, causing the resource server to be unable to distinguish an access
-token authorized by a resource owner from an access token authorized by a client
-itself, the authorization server SHOULD NOT allow clients to influence their
+identifiers, causing the resource server to be unable to distinguish whether an access
+token is meant to give access to end-user resources or resources for specific client,
```

Comment:
This evokes the same confusion as above.  Maybe reply in this GitHub comment thread giving examples of end-user resources and client resources to help us work out less confusing wording together.

In my (maybe overly simplistic view), resources are accessed by clients at resource servers, having been issued an access token by an authorization server, possibly with end-user approval.  Neither of the descriptions "end-user resources" or "resources for a specific client" map cleanly to this model, at least as I see it.

Reply:
The original concern that was raised was roughly the following:

In RFC9068 (JWT access token profile), the following is defined:

> `sub` REQUIRED - as defined in [Section 4.1.2](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.2) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)]. In cases of access tokens obtained through grants where a resource owner is involved, such as the authorization code grant, the value of "sub" SHOULD correspond to the subject identifier of the resource owner. In cases of access tokens obtained through grants where no resource owner is involved, such as the client credentials grant, the value of "sub" SHOULD correspond to an identifier the authorization server uses to indicate the client application.

This distinguishes between two cases - grants with involvement of the resource owner, and those without. In the first case, the subject value is an identifier for the resource owner, in the second case, for a client.

The potential attack is as follows: If a client can freely choose its client id, and therefore the `sub` value that ends up in the access token, this might lead to confusion at the resource server, who might accidentally give the client access to the resource owner's resources as both are indistinguishable from just looking at the `sub` value.

The current text in the BCP tries to describe this in a generic way, as there might be bespoke mechanisms out there that work similar to RFC9068 .

Reply:
The generic language didn't work for me, as I couldn't figure out what the issue was or how it could occur.  I also didn't realize that the cited text on `sub` from 9068 existed.

I suggest citing at least some of that text in the BCP and describing the problem/attack with respect to it.  Then if you want to say that there may be more general forms of this scenario, do so.

I'm trying to have us get this description to a state where normal readers familiar with OAuth will understand both the problem and how the proposed solution relates to it.  Thanks.

Reply:
I made a proposal based on the text from RFC9068, please re-review.